linkerd/proxy/http/src/detect.rs

```diff
@@ -0,0 +1,188 @@
+use crate::Version;
+use bytes::BytesMut;
+use linkerd2_error::Error;
+use linkerd2_io::{self as io, AsyncReadExt};
+use linkerd2_proxy_transport::Detect;
+use tracing::{debug, trace};
+
+const H2_PREFACE: &'static [u8] = b"PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
+
+#[derive(Clone, Debug, Default)]
+pub struct DetectHttp(());
+
+#[async_trait::async_trait]
+impl Detect for DetectHttp {
+    type Protocol = Version;
+
+    async fn detect<I: io::AsyncRead + Send + Unpin + 'static>(
+        &self,
+        io: &mut I,
+        buf: &mut BytesMut,
+    ) -> Result<Option<Version>, Error> {
+        let mut scan_idx = 0;
+        let mut maybe_h1 = true;
+        let mut maybe_h2 = true;
+
+        loop {
+            // Read data from the socket or timeout detection.
+            trace!(
+                capacity = buf.capacity(),
+                scan_idx,
+                maybe_h1,
+                maybe_h2,
+                "Reading"
+            );
+            let sz = io.read_buf(buf).await?;
+            if sz == 0 {
+                // No data was read because the socket closed or the
+                // buffer capacity was exhausted.
+                debug!(read = buf.len(), "Could not detect protocol");
+                return Ok(None);
+            }
+
+            // HTTP/2 checking is faster because it's a simple string match. If
+            // we have enough data, check it first. In almost all cases, the
+            // whole preface should be available from the first read.
+            if maybe_h2 {
+                if buf.len() < H2_PREFACE.len() {
+                    // Check the prefix we have already read to see if it looks likely to be HTTP/2.
+                    if buf[..] == H2_PREFACE[..buf.len()] {
+                        // If it looks like HTTP/2, it's not HTTP/1.
+                        maybe_h1 = false;
+                    } else {
+                        maybe_h2 = false;
+                    }
+                } else {
+                    trace!("Checking H2 preface");
+                    if &buf[..H2_PREFACE.len()] == H2_PREFACE {
+                        trace!("Matched HTTP/2 prefix");
+                        return Ok(Some(Version::H2));
+                    }
+
+                    // Not a match. Don't check for an HTTP/2 preface again.
+                    maybe_h2 = false;
+                }
+            }
+
+            if maybe_h1 {
+                // Scan up to the first line ending to determine whether the
+                // request is HTTP/1.1. HTTP expects \r\n, so we just look for
+                // any \n to indicate that we can make a determination.
+                if buf[scan_idx..].contains(&b'\n') {
```

Comment:
:+1: this is much simpler than what you had here yesterday --- and (IIRC) `contains`) should use a fast SIMD `memchr` impl when SIMD instructions are available.